packages/extras/src/lib/components/Sparkles/Sparkles.svelte

```diff
@@ -0,0 +1,102 @@
+<script lang="ts">
+	import { useFrame, Three  } from '@threlte/core';
+	import {
+		Vector2,
+		Vector3,
+		Vector4,
+		Color,
+		type ColorRepresentation,
+		ShaderMaterial
+	} from 'three';
+	import { randFloatSpread } from 'three/src/math/MathUtils';
+	import { SparkleShader } from './Shader';
+  import { BufferGeometry } from 'three'
+  import { BufferAttribute } from 'three'
+  import { Points } from 'three'
+	import type { SparklesProperties } from '../../types/components';
+
+	export let count: SparklesProperties['count'] = 100;
+	export let speed: SparklesProperties['speed'] = 1;
+	export let opacity: SparklesProperties['opacity'] = 1;
+	export let color: SparklesProperties['color'] = 100;
+	export let size: SparklesProperties['size']= Math.random();
+	export let scale: SparklesProperties['scale'] = 1;
+	export let noise: SparklesProperties['noise'] = 1;
```

Comment:
Currently only works with remount. Should I make them reactive.

Reply:
Should I use the implementation of useMemo we have?

Reply:
The implementation details are up to you as long as it works :) I guess memoization is not needed in this case, regular reactive statements will do just fine.

Reply:
Using `$: reactiveVar = someValue` doesn't work on initial render since it is undefined. When checking with if statements to see if the value is set then it will.

I've tried doing the following and that works:
```js
// All these variables are 'undefined' at initial render for some reason.
$: opacities = usePropAsIsOrAsAttribute<number>(count, opacity);
$: sizes = usePropAsIsOrAsAttribute<number>(count, size, Math.random);
$: speeds = usePropAsIsOrAsAttribute<number>(count, speed);
$: noises = usePropAsIsOrAsAttribute<typeof noise>(count * 3, noise);
$: colors = usePropAsIsOrAsAttribute<ColorRepresentation>(
			color === undefined ? count * 3 : count,
			!isFloat32Array(color) ? new Color(color) : color,
			() => 1
		)

$: {
   // Waiting to see if all the variables are set and then applying them to the geometry
   if(opacities && sizes && speeds && noises && colors) {
        geometry.setAttribute('position', new BufferAttribute(positions, 3))
        geometry.setAttribute('size', new BufferAttribute(sizes, 1))
	geometry.setAttribute('opacity', new BufferAttribute(opacities, 1))
	geometry.setAttribute('speed', new BufferAttribute(speeds, 1))
	geometry.setAttribute('color', new BufferAttribute(colors, 3))
	geometry.setAttribute('noise', new BufferAttribute(noises, 3))
   }
}
```

Even though it works, I feel like this isn't the right way of doing it. Do you happen to know a beter solution for this?

Reply:
If it's necessary to set all values, then yes, this approach is fine and valid. From my understanding you provide default values for all props in question, so the check if everything is set is redundant, right? I'd probably make the call to `geometry.setAttribute('name', new BufferAttribute(variable, 3))` a function:

```ts
const setAttribute = (name: string, value: any, size: number) => {
  geometry.setAttribute(name, new BufferAttribute(value, size))
}
```
and use it like this:

```ts
$: setAttribute('noises', usePropAsIsOrAsAttribute<typeof noise>(count * 3, noise), 3)
```

Would that suite the use case? Your call :)

Reply:
Nice!